### PR TITLE
[core] os.path -> Path

### LINKED ISSUE #4: Change os.path to Pathlib
Change `storage_json.py` file paths to use pathlib (and all other code paths using it)

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix?

This is a big migration from using `os.path` to using `Pathlib` in as many places as feasible.

Please test extensively as I am sure there will be some edge cases.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- closes https://github.com/esphome/backlog/issues/4

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jesserockz** [comment]
dangit, path.walk is python 3.12 :facepalm:

File: esphome/storage_json.py [hunk 1]
```diff
@@ -176,8 +177,8 @@ def from_wizard(
         )
 
     @staticmethod
-    def _load_impl(path: str) -> StorageJSON | None:
-        with codecs.open(path, "r", encoding="utf-8") as f_handle:
+    def _load_impl(path: Path) -> StorageJSON | None:
+        with codecs.open(str(path), "r", encoding="utf-8") as f_handle:
```
Comment (bdraco) [inline esphome/storage_json.py]: Can we use path.open here?

**Comment by bdraco** [comment]
Not sure how soon you want to merge this but I could spend some time this week improving the test coverage if you aren't in a hurry

**Comment by jesserockz** [comment]
After release is fine

File: esphome/storage_json.py (hunk 1 above)
Comment (bdraco) [reply esphome/storage_json.py]: a1676b4ab0f61198200eb5d786d13696699194ac

**Comment by bdraco** [comment]
Still 73 missing lines.

next PR to add cover https://github.com/esphome/esphome/pull/10669

**Comment by bdraco** [comment]
69 lines left. It looks like the symlink resolution might be a bit different in some places but that seems ok

**Comment by bdraco** [comment]
61 lines left to cover

**Comment by bdraco** [comment]
more dashboard tests https://github.com/esphome/esphome/pull/10682

**Comment by bdraco** [comment]
```
        build_path = storage.get("build_path")
        firmware_bin_path = storage.get("firmware_bin_path")
```

This is a problem. It needs to be converted from str to Path when loaded from json

Will fix it in the next push

**Comment by bdraco** [comment]
dashboard has some issues still
```
2025-09-11 16:28:48,349 ERROR Exception in callback functools.partial(<function WebSocketProtocol._run_callback.<locals>.<lambda> at 0x1064e6340>, <Task finished name='Task-48' coro=<EsphomeCommandWebSocket.on_message() done, defined at /Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py:200> exception=TypeError("expected string or bytes-like object, got 'PosixPath'")>)
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 668, in <lambda>
    self.stream.io_loop.add_future(result, lambda f: f.result())
                                                     ~~~~~~~~^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 214, in on_message
    await handlers[type_](self, json_message)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 222, in handle_spawn
    _LOGGER.info("Running command '%s'", " ".join(shlex_quote(x) for x in command))
                                         ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 222, in <genexpr>
    _LOGGER.info("Running command '%s'", " ".join(shlex_quote(x) for x in command))
                                                  ~~~~~~~~~~~^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/util.py", line 91, in shlex_quote
    if re.search(r"[^\w@%+=:,./-]", s) is None:
       ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/re/__init__.py", line 177, in search
    return _compile(pattern, flags).search(string)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
TypeError: expected string or bytes-like object, got 'PosixPath'
2025-09-11 16:28:48,352 ERROR Uncaught exception GET /compile (127.0.0.1)
HTTPServerRequest(protocol='http', host='0.0.0.0:6052', method='GET', uri='/compile', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/web.py", line 1848, in _execute
    result = await result
             ^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 277, in get
    await self.ws_connection.accept_connection(self)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 890, in accept_connection
    await self._accept_connection(handler)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 973, in _accept_connection
    await self._receive_frame_loop()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 1127, in _receive_frame_loop
    await self._receive_frame()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 1218, in _receive_frame
    await handled_future
TypeError: expected string or bytes-like object, got 'PosixPath'
```

**Comment by bdraco** [comment]
Change in behavior on the dashboard, its now finding all files in the esphome dir when I do `esphome dashboard .`

File: esphome/util.py [hunk 2]
```diff
@@ -272,25 +271,28 @@ def __repr__(self):
...
     for config in configs:
-        if os.path.isfile(config):
+        config = Path(config)
+        if not config.exists():
+            raise FileNotFoundError(f"Config path '{config}' does not exist!")
+        if config.is_file():
             files.append(config)
         else:
-            files.extend(os.path.join(config, p) for p in os.listdir(config))
+            files.extend(config.glob("**/*"))
```
Comment (bdraco) [inline esphome/util.py]: This is where the regression comes from

**Comment by bdraco** [comment]
full trace for the dashboard regression
```
2025-09-11 16:33:32,904 ERROR Exception in callback functools.partial(<function WebSocketProtocol._run_callback.<locals>.<lambda> at 0x103ea8ea0>, <Task finished name='Task-37' coro=<EsphomeCommandWebSocket.on_message() done, defined at /Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py:200> exception=TypeError("expected string or bytes-like object, got 'PosixPath'")>)
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/ioloop.py", line 758, in _run_callback
    ret = callback()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 668, in <lambda>
    self.stream.io_loop.add_future(result, lambda f: f.result())
                                                     ~~~~~~~~^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 214, in on_message
    await handlers[type_](self, json_message)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 222, in handle_spawn
    _LOGGER.info("Running command '%s'", " ".join(shlex_quote(x) for x in command))
                                         ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/dashboard/web_server.py", line 222, in <genexpr>
    _LOGGER.info("Running command '%s'", " ".join(shlex_quote(x) for x in command))
                                                  ~~~~~~~~~~~^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/esphome/util.py", line 91, in shlex_quote
    if re.search(r"[^\w@%+=:,./-]", s) is None:
       ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/re/__init__.py", line 177, in search
    return _compile(pattern, flags).search(string)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
TypeError: expected string or bytes-like object, got 'PosixPath'
2025-09-11 16:33:32,907 ERROR Uncaught exception GET /compile (127.0.0.1)
HTTPServerRequest(protocol='http', host='0.0.0.0:6052', method='GET', uri='/compile', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/web.py", line 1848, in _execute
    result = await result
             ^^^^^^^^^^^^
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 277, in get
    await self.ws_connection.accept_connection(self)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 890, in accept_connection
    await self._accept_connection(handler)
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 973, in _accept_connection
    await self._receive_frame_loop()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 1127, in _receive_frame_loop
    await self._receive_frame()
  File "/Users/bdraco/home-assistant/venv/lib/python3.13/site-packages/tornado/websocket.py", line 1218, in _receive_frame
    await handled_future
TypeError: expected string or bytes-like object, got 'PosixPath'
```

fixed in ae4eebad1

File: esphome/util.py (hunk 2 above)
Comment (bdraco) [reply esphome/util.py]: fixed in cb1b19f3d0409c61d2b628ddd318dedaf2033653

**Comment by bdraco** [comment]
Added more coverage for util since we had regressions there https://github.com/esphome/esphome/pull/10683

**Comment by bdraco** [comment]
down to 50 lines

**Comment by bdraco** [comment]
https://github.com/esphome/esphome/pull/10688 should cover another 12 lines

**Comment by bdraco** [comment]
Down to 37 lines. Did find some actual problems so probably good to keep going

**Comment by bdraco** [comment]
next one https://github.com/esphome/esphome/pull/10700

**Comment by bdraco** [comment]
down to 23 lines uncovered

**Comment by bdraco** [comment]
mostly the only one I'm worried about left is `add_includes`

**Comment by bdraco** [comment]
https://github.com/esphome/esphome/pull/10723 is the next round. Bit more messy because we have to account for windows

**Comment by bdraco** [comment]
> mostly the only one I'm worried about left is `add_includes`

Looks like there was a regression. all sorted

**Comment by bdraco** [comment]
Down to 18 lines.  I think we should be ok here now. Hopefully didn't miss anything

**Review by bdraco — DISMISSED** [review]

**Review by bdraco — DISMISSED** [review]

**Comment by bdraco** [comment]
I keep fixing conflicts, and than I make more 😢

**Comment by bdraco** [comment]
Down to 15 lines

**Comment by bdraco** [comment]
Will fix conflicts later tonight or tomorrow if I run out of time

**Review by bdraco — APPROVED** [review]

**Comment by balloob** [comment]
This broke `esphome run`, fixed here https://github.com/esphome/esphome/pull/10807

**Comment by bdraco** [comment]
Good thing we held off and waited to merge at top of cycle
